bugsnag-plugin-android-ndk/src/main/jni/bugsnag_ndk.c

```diff
@@ -157,30 +157,37 @@ Java_com_bugsnag_android_ndk_NativeBridge_deliverReportAtPath(
       jclass interface_class =
           bsg_safe_find_class(env, "com/bugsnag/android/NativeInterface");
       if (interface_class == NULL) {
-        return;
+        goto exit;
       }
 
       // lookup NativeInterface.deliverReport()
       jmethodID jdeliver_method = bsg_safe_get_static_method_id(
           env, interface_class, "deliverReport", "([B[BLjava/lang/String;)V");
       if (jdeliver_method == NULL) {
-        return;
+        goto exit;
       }
 
-      size_t payload_length = bsg_strlen(payload);
-      jbyteArray jpayload = (*env)->NewByteArray(env, payload_length);
-      (*env)->SetByteArrayRegion(env, jpayload, 0, payload_length,
-                                 (jbyte *)payload);
+      // generate payload bytearray
+      jbyteArray jpayload = bsg_byte_ary_from_string(env, payload);
+      if (jpayload == NULL) {
+        goto exit;
+      }
 
-      size_t stage_length = bsg_strlen(event->app.release_stage);
-      jbyteArray jstage = (*env)->NewByteArray(env, stage_length);
-      (*env)->SetByteArrayRegion(env, jstage, 0, stage_length,
-                                 (jbyte *)event->app.release_stage);
+      // generate releaseStage bytearray
+      jbyteArray jstage =
+          bsg_byte_ary_from_string(env, event->app.release_stage);
+      if (jstage == NULL) {
+        goto exit;
+      }
 
-      jstring japi_key = (*env)->NewStringUTF(env, event->api_key);
-      (*env)->CallStaticVoidMethod(env, interface_class, jdeliver_method,
-                                   jstage, jpayload, japi_key);
+      // call NativeInterface.deliverReport()
+      jstring japi_key = bsg_safe_new_string_utf(env, event->api_key);
+      if (japi_key != NULL) {
+        (*env)->CallStaticVoidMethod(env, interface_class, jdeliver_method,
+                                     jstage, jpayload, japi_key);
+      }
       (*env)->DeleteLocalRef(env, japi_key);
+
       (*env)->ReleaseByteArrayElements(env, jpayload, (jbyte *)payload,
```

Comment:
These should also be in the cleanup code at the end. They could be live objects when the code fails later in the function.